src/vs/workbench/services/lifecycle/test/electron-sandbox/lifecycleService.test.ts

```diff
@@ -155,5 +156,32 @@ suite('Lifecycleservice', function () {
 		assert.strictEqual(joinCalled, true);
 	});
 
+	test('onWillShutdown - join order', async function () {
+		const order: string[] = [];
+
+		disposables.add(lifecycleService.onWillShutdown(e => {
+			e.join(async () => {
+				order.push('disconnect start');
+				await timeout(1);
+				order.push('disconnect end');
+			}, { id: 'test', label: 'test', order: WillShutdownJoinerOrder.Last });
+
+			e.join((async () => {
+				order.push('default start');
+				await timeout(1);
```

Comment:
If we are using `timeout` I think we should also use fake timers to reduce flakyness.

Reply:
This is more just to have some async to verify that the code waits for the promise to resolve in the first stage before starting the last stage. If this does flake that means there is a bug in the implementation: the order of these events should be deterministic.